_posts/2019-08-26-how-to-use-gruntwork-service-catalog.adoc

```diff
@@ -0,0 +1,1934 @@
+
+* You're always deploying "latest" from a branch. No versioning or easy rollback.
+* Lots of command-line arguments to pass. Easy to make mistakes. Most teams end up creating hacky wrapper scripts.
+* Lots of backend configuration to copy for each module. Manually setting a unique `key` for each module is repetitive
+  and error prone.
+* From a quick glance at the code, it's not clear what accounts, environments, or regions you deploy to. Figuring this
+  out requires digging through every `.tfvars` file.
+
+[[deploy_using_terragrunt]]
+===== Deploy using Terragrunt
+
+Another option is to use https://github.com/gruntwork-io/terragrunt[Terragrunt], an open source wrapper for Terraform
+that helps with some of the drawbacks mentioned in the previous approach.
+
+The first step with Terragrunt is to version your code. You can do this by creating Git tags in
+`infrastructure-modules`:
+
+----
+cd infrastructure-modules
+git tag -a "v0.0.1" -m "Created vpc-app module"
+git push --follow-tags
+----
+
+This will allow you to deploy different versions of your module in different environments (e.g., `v0.0.1` in prod and
+`v0.0.2` in stage) and rollback to previous versions if necessary. With Terragrunt, we recommend defining your live
+environments in a separate repo called `infrastructure-live` that uses a folder structure with the following format:
+
+----
+infrastructure-live
+  └ <account>
+    └ terragrunt.hcl
+    └ _global
+    └ <region>
+      └ _global
+      └ <environment>
+        └ <resource>
+          └ terragrunt.hcl
+----
+
+Where:
+
+<account>::
+  A the top level, you have accounts (e.g., an AWS account) or projects (e.g., GCP project).
+
+<region>::
+  Within each account, there will be one or more regions (e.g., in AWS, `us-east-1`, `eu-west-1`, etc). There may also
+  be a `_global` folder that defines resources that are available across all the regions in this account, such as
+  IAM users and DNS settings. Each account also has a root `terragrunt.hcl` file that defines common Terraform settings
+  that apply to the entire account, such as what backend to use to store Terraform state.
+
+<environment>::
+  Within each region, there will be one or more environments, such as dev, stage, prod, mgmt, etc. There may also be a
+  `_global` folder that defines resources that are available across all the environments in this region.
+
+<resource>::
+  Within each environment, you use Terraform modules to deploy one or more resources, such as servers, databases load
+  balancers, and so on. Each module is configured via a `terragrunt.hcl` file.
+
+For example, if you were using AWS, with separate accounts for staging and production (see
+<<production_grade_aws_account_structure>>), and you wanted to deploy the `vpc-app` module in the `us-east-2` region in
+each of these accounts, the folder structure would look like this:
+
+----
+infrastructure-live
+  └ staging
+    └ terragrunt.hcl
+    └ us-east-2
+      └ stage
```

Comment:
If one uses different accounts for different environments, this level here looks very redundant. Is there a way we can clarify on this a bit more?

Reply:
Updated with a NOTE to explain it: https://github.com/gruntwork-io/gruntwork-io.github.io/pull/151/commits/4603b48a336569db90352172790bc29347291c75